### ISSUE TITLE
Feature request: typeless variables

### ISSUE THREAD

**Issue by stugol** [ozra/onyx-lang#27]
Essentially variants. Obviously they would incur a performance hit, but sometimes ease-of-use trumps execution speed. Use of a variant would never trigger a compilation error:
```
type T
	a-fn() ->

v = T()
puts v.some-method     # run-time error, not compilation error
```
Note that I am not talking about unioned types. I mean _true_ typeless variables. Essentially Ruby variables.

**Comment by ozra** [comment]
Why would you ever want that? One of the few things that studies seem to agree upon is that untyped programming _is less productive_.

**Comment by stugol** [comment]
Crystal's type-unioning can result in a hideous mess. I've literally had types like this:
```
Array(Int32) | Array(Int64) | Array(UInt32) | Array(Int32|Int64) | Array(Int32|Int64|UInt32) | Array(UInt32|Int32) | Nil
```
Not to mention the nightmare of storing more than one damn type of thing in an array:
```
do-something(value String) ->

a = ["fred", 3, true]
do-something a[0]			-- compile error
```
This would be perfectly acceptable in Ruby.

Maybe instead of making something completely untyped, have some way to have a variable _automatically cast_ to any type in its union. For example:
```
a = ["fred", 3, true]
do-something a[0]			-- equivalent to do-something a[0] as String
```
We'd probably want some way to specify whether we wanted strict or relaxed typing, but you get the idea.

**Comment by ozra** [comment]
Of course, such mixing is not too common in "performance code" (which is why we use an imperative design - otherwise, if you don't need the sheer power, you're better of coding in a functional language)

Nevertheless: As you mention, automatic coercion in certain situations is the much more reasonable solution to this then weakening the type system.
C++ does this in a lot, and chained ways, at times.

Personally I think a symbol should be used for this "fix this for me", so instead of explicit casting, you just "allow casting". This is something I've been tossing around for about a year. The concept jugglings involves more constructs and I have still not arrived at a really clean solution which is why it's not posted.

I think, after getting up to speed on current issues, I'll put together an "overview issue" relating to value/type/qualification modifying and that nasty 4D stuff.

**Comment by stugol** [comment]
Maybe a use for the unary `*`?

**Comment by ozra** [comment]
Closed in favour of #55
